File: lib/feature/payment/infrastructure/persistence/payment_repository_impl.dart
import 'dart:convert';
import 'package:dartz/dartz.dart';

import '../../../../shared/model/response_model.dart';
import '../../domain/model/response/payment_category.dart';
import '../../domain/model/response/payment_method.dart';
import '../../domain/repository/payment_repository.dart';
import '../data_source/payment_remote_data_source.dart';

class PaymentRepositoryImpl implements PaymentRepository {
  final PaymentRemoteDataSource remoteDataSource;

  PaymentRepositoryImpl({PaymentRemoteDataSource? remoteDataSource})
      : remoteDataSource = remoteDataSource ?? PaymentRemoteDataSource();

  @override
  Future<Either<ResponseModel, List<PaymentCategory>>>
      getPaymentCategory() async {
    final response = await remoteDataSource.getPaymentCategory();
    if (!response.isError) {
      try {
        final body = jsonDecode(response.result?.body ?? '');
        final data = body['data'];
        return Right(paymentCategoryFromJson(data));
      } catch (e) {
        return Left(ResponseModel(isError: true, result: response.result, msg: e.toString()));
      }
    }
    return Left(response);
  }

  @override
  Future<Either<ResponseModel, List<PaymentResponse>>> getPaymentMethod({
    String? from,
    String? categoriesKey,
  }) async {
    final response = await remoteDataSource.getPaymentMethod(
      from: from,
      categoriesKey: categoriesKey,
    );
    if (!response.isError) {
      try {
        final body = jsonDecode(response.result?.body ?? '');
        final data = body['data'];
        if (data is List) {
          return Right(
            List<PaymentResponse>.from(
              data.map((e) => PaymentResponse.fromMap(e)),
            ),
          );
        }
        return const Right([]);
      } catch (e) {
        return Left(ResponseModel(isError: true, result: response.result, msg: e.toString()));
      }
    }
    return Left(response);
  }

  @override
  Future<Either<ResponseModel, PaymentResponse>>
      getDetailPaymentMethod() async {
    final response = await remoteDataSource.getDetailPaymentMethod();
    if (!response.isError) {
      try {
        final body = jsonDecode(response.result?.body ?? '');
        final data = body['data'];
        if (data != null) {
          return Right(PaymentResponse.fromMap(data));
        }
      } catch (e) {
        return Left(ResponseModel(isError: true, result: response.result, msg: e.toString()));
      }
    }
    return Left(response);
  }
}

File: src/features/payment/infrastructure/persistence/payment_repository_impl.ts
import { Either, left, right } from "@/shared/utils/utility/either";
import { ResponseModel } from "@/shared/domain/model/response_model";
import { PaymentRepository } from "../../domain/repository/payment_repository";
import { PaymentRemoteDataSource } from "../data_source/remote/payment_remote_data_source";
import { PaymentCategoryResponse } from "../../domain/model/response/payment_category_response";
import { PaymentMethodResponse } from "../../domain/model/response/payment_method_response";

export class PaymentRepositoryImpl implements PaymentRepository {
  private dataSource: PaymentRemoteDataSource;

  constructor(dataSource?: PaymentRemoteDataSource) {
    this.dataSource = dataSource ?? new PaymentRemoteDataSource();
  }

  async getPaymentCategory(token?: string): Promise<Either<ResponseModel, PaymentCategoryResponse[]>> {
    const res = await this.dataSource.getPaymentCategory(token);
    if (res.status && Array.isArray(res.data)) {
      return right(res.data);
    }
    return left(res);
  }

  async getPaymentMethod(
    params?: {
      paymentGatewayKey?: string;
      paymentGatewayId?: string;
      categoriesKey?: string;
    },
    token?: string,
  ): Promise<Either<ResponseModel, PaymentMethodResponse[]>> {
    const res = await this.dataSource.getPaymentMethod(params, token);
    if (res.status && Array.isArray(res.data)) {
      return right(res.data);
    }
    return left(res);
  }

  async getDetailPaymentMethod(key?: string, token?: string): Promise<Either<ResponseModel, PaymentMethodResponse>> {
    const res = await this.dataSource.getDetailPaymentMethod(key, token);
    if (res.status && res.data) {
      return right(res.data);
    }
    return left(res);
  }
}
